phase1-coordinator/src/coordinator_state.rs

```diff
@@ -916,6 +966,190 @@ impl CoordinatorState {
         }
     }
 
+    /// Reset the progress of the current round, back to how it was in
+    /// its initialized state, however this does maintain the drop
+    /// status of participants.
+    ///
+    /// If `force_rollback` is set to `true` the coordinator will be
+    /// forced to reset to the end of the previous round and begin
+    /// waiting for new participants again before restarting the
+    /// current round. If set to `false` the rollback will only occur
+    /// if there are no available contributors or no available
+    /// verifiers left in the round.
+    ///
+    /// Returns [CoordinatorError::RoundDoesNotExist] if
+    /// [CoordinatorState::current_round_height] is set to `None`.
+    ///
+    /// Returns [CoordinatorError::RoundHeightIsZero] if
+    /// [CoordinatorState::current_round_height] is set to `Some(0)`.
+    pub fn reset_current_round(
+        &mut self,
+        force_rollback: bool,
+        time: &dyn TimeSource,
+    ) -> Result<ResetCurrentRoundStorageAction, CoordinatorError> {
+        let span = tracing::error_span!("reset_round", round = self.current_round_height.unwrap_or(0));
+        let _guard = span.enter();
+
+        let current_round_height = self.current_round_height.ok_or(CoordinatorError::RoundDoesNotExist)?;
+        if current_round_height == 0 {
+            return Err(CoordinatorError::RoundHeightIsZero);
+        }
+
+        tracing::warn!("Resetting round {}.", current_round_height);
+
+        let finished_contributors = self
+            .finished_contributors
+            .get(&current_round_height)
+            .cloned()
+            .unwrap_or_else(|| HashMap::new());
+
+        let finished_verifiers = self
+            .finished_verifiers
+            .get(&current_round_height)
+            .cloned()
+            .unwrap_or_else(|| HashMap::new());
+
+        let number_of_contributors = self.current_contributors.len() + finished_contributors.len();
+        let number_of_verifiers = self.current_verifiers.len() + finished_verifiers.len();
+        let number_of_chunks = self.environment.number_of_chunks() as u64;
+
+        let current_contributors = self
+            .current_contributors
+            .clone()
+            .into_iter()
+            .chain(finished_contributors.clone().into_iter())
+            .enumerate()
+            .map(|(bucket_index, (participant, mut participant_info))| {
+                let bucket_id = bucket_index as u64;
+                let tasks = initialize_tasks(bucket_id, number_of_chunks, number_of_contributors as u64)?;
+                participant_info.restart_tasks(tasks, time)?;
```

Comment:
It looks like one of the key parts of the change, is it? To reset the state of all contributors who is still in the round to the initial state similar to what we have on round start.

Reply:
yep that's right